spark/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala

```diff
@@ -547,25 +574,273 @@ object Checkpoints extends DeltaLogging {
+   * create the same files.
+   */
+  def getCheckpointWritePath(
+      conf: Configuration,
+      logPath: String,
+      version: Long,
+      numParts: Int,
+      part: Int,
+      useRename: Boolean,
+      v2CheckpointEnabled: Boolean): (Path, Path) = {
+    def getCheckpointWritePath(path: Path): Path = {
+      if (useRename) {
+        val tempPath =
+          new Path(path.getParent, s".${path.getName}.${UUID.randomUUID}.tmp")
+        DeltaFileOperations.registerTempFileDeletionTaskFailureListener(conf, tempPath)
+        tempPath
+      } else {
+        path
+      }
+    }
+    val logPathObj = new Path(logPath)
+    val destinationName: Path = if (v2CheckpointEnabled) {
+      newV2CheckpointSidecarFile(logPathObj, version, numParts, part + 1)
+    } else {
+      if (numParts > 1) {
+        assert(part < numParts, s"Asked to create part: $part of max $numParts in checkpoint.")
+        checkpointFileWithParts(logPathObj, version, numParts)(part)
+      } else {
+        checkpointFileSingular(logPathObj, version)
+      }
+    }
+
+    getCheckpointWritePath(destinationName) -> destinationName
+  }
+
+  /**
+   * Writes a top-level V2 Checkpoint file which may point to multiple
+   * sidecar files.
+   *
+   * @return A tuple containing
+   *          1. [[FileStatus]] of the newly created top-level V2Checkpoint.
+   *          2. The sequence of actions that were written to the top-level file.
+   *          3. An instance of the LastCheckpointV2 containing V2-checkpoint related
+   *           metadata which can later be written to LAST_CHECKPOINT
+   *          4. Schema of the newly written top-level file (only for parquet files)
+   */
+  protected[delta] def writeTopLevelV2Checkpoint(
+      v2CheckpointFormatOpt: Option[V2Checkpoint.Format],
+      sidecarCheckpointFiles: Array[SerializableFileStatus],
+      spark: SparkSession,
+      schema: StructType,
+      snapshot: Snapshot,
+      deltaLog: DeltaLog,
+      rowsWrittenInCheckpointJob: Long,
+      parquetFilesSizeInBytes: Long,
+      hadoopConf: Configuration,
+      useRename: Boolean) : (FileStatus, Seq[Action], LastCheckpointV2, Option[StructType]) = {
+    val v2CheckpointFormat = v2CheckpointFormatOpt.get
+    // Write the main v2 checkpoint file.
+    val sidecarFilesWritten = sidecarCheckpointFiles.map(SidecarFile(_)).toSeq
+    val sidecarFileSchemaOpt =
+      Checkpoints.checkpointSchemaToWriteInLastCheckpointFile(spark, schema)
+    val checkpointMetadata = CheckpointMetadata(
+      version = snapshot.version,
+      sidecarNumActions = rowsWrittenInCheckpointJob,
+      sidecarSizeInBytes = parquetFilesSizeInBytes,
+      numOfAddFiles = snapshot.numOfFiles,
+      sidecarFileSchemaOpt = sidecarFileSchemaOpt)
+    val nonFileActionsToWrite =
+      (checkpointMetadata +: sidecarFilesWritten) ++ snapshot.nonFileActions
+    // By default, create v2 checkpoints in json format with fileactions in sidecar files
```

Comment:
This comment doesn't seem to be at the right place? Here we are not relying on the default - we are relying on variable `v2CheckpointFormat`.

Reply:
Line 689 is wrong ( val v2CheckpointFormat = v2CheckpointFormatOpt.get). The checkpoint v2 format is optional. I should probably move this comment above line 689 and change line 689 to 
```
val v2CheckpointFormat = v2CheckpointFormatOpt.getOrElse( V2Checkpoint.Format.JSON)

Reply:
Never mind, I forgot that this logic is already present where before `writeTopLevelV2Checkpoint` is called.

Reply:
The getOrElse is already taken care of while initializing v2CheckpointFormatOpt - https://github.com/delta-io/delta/pull/2031/files#diff-ffbd94fa1932ccd2f2238ca3ef7911f36c92bfa5bcaf38e3bf43e9fe012e1625R431 .

Also `v2CheckpointEnabled` is defined using  `v2CheckpointFormatOpt` only. Based on this, v2CheckpointFormatOpt must be non-empty here.